massgov/mayflower-patternlab
# DP-8339: Automate Mayflower Release Process

Any PRs being created needs a changelog.txt file before being merged into dev. See: [Change Log Instructions](https://github.com/massgov/mayflower/blob/dev/docs/change-log-instructions.md)


## Description

The high level goal of this ticket is to make Mayflower releases much simpler.  We should be able to trigger all of the release steps (NPM, S3, Artifacts) just by pushing a tag to this repository, and letting CircleCI do the work.  This PR implements this automation.

Successful Build: https://circleci.com/workflow-run/d3b50893-cee6-4eb1-94e0-6cce5dcfd7a7
This build released the 5.17.0-alpha1 tag to the mayflower-artifacts repo and NPM (see https://www.npmjs.com/package/@massds/mayflower).  It didn't update the root S3 installation because it's not a stable tag (only stable tags are released to S3's / directory).

 Note that Visual Regression testing shows **0** differences here, so we're really only concerned with how this integrates with other applications.  There ended up being no template or CSS changes, and only one minor JS change required, so as long as it works with Drupal and Mayflower React, this should be a no-impact change, despite the intimidating list below. 

### Changes:
* Updates to all build-only dependencies (didn't touch neat, bourbon, or babel) including Gulp 4.
* Refactor gulp tasks on Gulp 4 and split into 3 distinct "task registries".
* Release built Pattern Lab display for stable tags to /v/${MAJOR_VERSION}
* Release built Pattern Lab display for branches to /b/${BRANCH}
* Exclude moment JS from custom JS bundle (saves 100k!).
* Additional minification for vendor JS bundle.
* Cleans mayflower-artifacts so it actually removes files that are removed on the source side.  This removes a lot of [old leftovers](https://github.com/massgov/mayflower-artifacts/compare/DP-8339-mayflower_release_revamp)
* Move to babel-preset-env, which is configured in package.json.
* Use a .browserslistrc for determining the list of supported browsers for autoprefixer.  We can eventually use this for babel, too, but need to update some of our code to work with newer versions of babel.
* Change most asset imports to be relative (should only affect the _meta templates).
* Determine the URL from an environment variable in Pattern Lab.
* Determine the Release from package.json in Pattern Lab.
* Determine the Release date from the date of the last commit in Pattern Lab.

### New Gulp tasks:

#### Pattern Lab
* `patternlab:build` - Compile Pattern Lab into `styleguide/public`
* `patternlab:watch` - Compile and start a live development server.
* `patternlab:publish` - Compile and deploy to the S3 bucket at /b/${BRANCH}.
* `patternlab:release` - Compile and deploy to the S3 bucket at /b/${BRANCH}.  If tagged, and tag is not a pre-release tag, also deploys to /v/${MAJOR_VERSION}, and /.

#### Artifacts
* `artifacts:build` - Clones artifact repo, cleans it, and builds assets into `styleguide/artifacts`
* `artifacts:publish` - Runs `artifacts:build`, then pushes to `$BRANCH` on the artifacts repository.
* `artifacts:release` - Runs `artifacts:build`, then pushes to `$BRANCH` and `$TAG` on the artifacts repository.

#### NPM
* `npm:build` - Builds the NPM package to `styleguide/npm`.
* `npm:release` - Runs `npm:build` and then publishes to the NPM registry.


## Open Questions:
* Do we still use the `svg2twig` and `svg-sprite` Gulp tasks?


### Followups needed:
* Remove old `/scripts` and Phing build after this work has been proven reliable.
* Automated cleanup of S3 bucket.
* Update Mayflower to latest babel and use `babel-preset-env` with `.browserslistrc`
* Review README.md and developer-deploy.md and make sure they still make sense.

### Once this is merged:
* Announce to #massgov:
    > Attention Mayflower developers: we have updated the way Mayflower is generated locally and in CircleCI.  Please be aware of the following changes.
    * You should run `npm install` before beginning work or you will get errors running Gulp.
    * Pre-release tags are no longer necessary for previewing changes in Drupal.  You can now preview a branch instead of needing a tag.  To do this, run the following command from the Drupal repository after pushing your changes to a Mayflower branch: `composer require palantirnet/mayflower:dev-DP-1000-my-branch`, where "DP-1000-my-branch" is the name of your branch.
    * The Mayflower release process is now handled mostly by CircleCI.  If you need to do a Mayflower release, please read the updated release instructions (https://github.com/massgov/mayflower/blob/dev/docs/release.md).



## Related Issue / Ticket

- [JIRA issue](https://jira.state.ma.us/browse/DP-8339)

## Steps to Test


1. Code review is probably the most important here.
2. Try setting up for local development and make sure you can still do all the same things.
2. Check the [deployed Pattern Lab instance](https://mayflower.digital.mass.gov/b/DP-8339-mayflower_release_revamp/index.html)
3. Check the [artifacts release tag](https://github.com/massgov/mayflower-artifacts/releases/tag/5.17.0-alpha1).  This is what will be pulled into Drupal.
4. Check the [NPM package](https://www.npmjs.com/package/@massds/mayflower). This is what will be pulled into Mayflower React.

## Screenshots
Use something like [licecap](http://www.cockos.com/licecap/) to capture gifs to demonstrate behaviors.


## Additional Notes:

Anything else to add?

#### Impacted Areas in Application


* 

#### @TODO


*

#### Today I learned...
